presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -418,15 +419,21 @@ private RowExpression pullExpressionThroughVariables(RowExpression expression, C
             for (RowExpression conjunct : new EqualityInference.Builder(functionManger).nonInferableConjuncts(expression)) {
                 if (determinismEvaluator.isDeterministic(conjunct)) {
                     RowExpression rewritten = equalityInference.rewriteExpression(conjunct, in(variables));
-                    if (rewritten != null) {
+                    if (rewritten != null && (hasVariableReferences(rewritten) || rewritten.equals(FALSE_CONSTANT))) {
                         effectiveConjuncts.add(rewritten);
                     }
+                    // If equality inference has reduced the predicate to an expression referring to only constants, it does not make sense to pull this predicate up
```

Comment:
Bit confused here, as to why do we have the condition to skip if rewritten has variable references? What if the variable reference is pointing to a constant in some project below?

Reply:
If `rewritten` has variable references that are `in(variables)` we do *not* skip them and pull them up (by adding them to `effectiveConjuncts`. The problem here is that `EqaulityInference` is designed to rewrite expressions to just constants (see [test](https://github.com/prestodb/presto/blob/185876fab2f8aa2ff5caf2a9da6284a9c227c9c4/presto-main/src/test/java/com/facebook/presto/sql/planner/TestEqualityInference.java#L302-L303)), even though, technically this expression does not refer to any `variables`

However, if the `rewritten` expression has no variable references, le.g `mod(10,5)=1` - this predicate should not be pulled up, to say, a Join node, because it has no references in the outputs of the Join and cannot impact any predicate inferencing there